### PR TITLE
Fix: API issue preventing multi-field updates

### PR DESCRIPTION
**Pull request by MyvTsv**
## Checklist before requesting a review

*Please delete options that are not relevant.*

- [x] I have performed a self-review of my code.

## Description

- It fixes !35540
- It fixes the issue that prevented modifying a field via the API when a field and a new tab were added to the same object.

### CODE REVIEW

File: inc/container.class.php [hunk 1]
```diff
@@ -1635,19 +1635,19 @@ public static function preItemUpdate(CommonDBTM $item)
...
-            $container = new self();
-            if (
-                count($data) == 0
-                || $container->updateFieldsValues($data, $item->getType(), isset($_REQUEST['massiveaction']))
-            ) {
-                $item->input['date_mod'] = $_SESSION['glpi_currenttime'];
-
-                return true;
+            foreach ($item->input['_plugin_fields_data'] as $containerClass) {
+                $data = $containerClass;
```
Comment (Rom1-B) [inline inc/container.class.php:1639]:
```suggestion
            foreach ($item->input['_plugin_fields_data'] as $data) {
```

File: inc/container.class.php [hunk 2]
```diff
@@ -1663,9 +1663,14 @@ public static function preItemUpdate(CommonDBTM $item)
      */
     public static function preItem(CommonDBTM $item)
     {
+        /** @var DBmysql $DB */
+        global $DB;
+
+        $c_id = [];
+        $loc_c = new PluginFieldsContainer();
```
Comment (Rom1-B) [inline inc/container.class.php:1670]: Please use explicit variable names

**Review by Rom1-B — CHANGES_REQUESTED** [review]

File: inc/container.class.php [hunk 3]
```diff
@@ -1635,19 +1635,19 @@ public static function preItemUpdate(CommonDBTM $item)
...
-                return true;
+            foreach ($item->input['_plugin_fields_data'] as $containerClass) {
+                $data = $containerClass;
+                //update data
+                $container = new self();
+                if (
+                    count($data) == 0
+                    || $container->updateFieldsValues($data, $item->getType(), isset($_REQUEST['massiveaction']))
+                ) {
+                    $item->input['date_mod'] = $_SESSION['glpi_currenttime'];
```
Comment (stonebuzz) [inline inc/container.class.php:1646]: You can `exit` from `foreach` here, no need to check for all container if one of them already trigger an update on `date_mod`

File: inc/container.class.php [hunk 4]
```diff
@@ -1675,57 +1680,56 @@ public static function preItem(CommonDBTM $item)
...
-                    return false;
+            foreach ($item->input as $key => $value) {
+                $container_id = self::findContainer(get_class($item), $type, $subtype);
+                $container_is_find = $loc_c->getFromDB($container_id);
+                if (!$DB->fieldExists(static::getTable(), $key) || !$container_is_find) {
+                    $container_id = self::findContainer(get_class($item));
+                    if ($container_id === false) {
+                        return false;
+                    }
+                }
```
Comment (stonebuzz) [inline inc/container.class.php:1691]:
```suggestion
                if (!$DB->fieldExists(static::getTable(), $key) || false === ($c_id = self::findContainer(get_Class($item), $type, $subtype))) {
                    // tries for 'tab'
                    if (false === ($c_id = self::findContainer(get_Class($item)))) {
                        return false;
                    }
                }
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

**Comment by stonebuzz** [comment]
@MyvTsv can you adapt `CHANGELOG.md`

**Review by Rom1-B — CHANGES_REQUESTED** [review]
waiting for customer feedback

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
+## [1.21.16] - yyyy-mm-dd
+
+### Fixed
+
+- Fix issue preventing field modification via API when a field and a new tab are added to the same object
```
Comment (stonebuzz) [inline CHANGELOG.md:12]:
```suggestion
- Fixed a problem with fields being updated if there are several containers on the object.
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

**Comment by MyvTsv** [comment]
https://github.com/pluginsGLPI/fields/pull/875

**Pull request closed without merging**